src/io/aleph/dirigiste/Pool.java

```diff
@@ -285,6 +282,109 @@ private void addObject(K key) {
         }
     }
 
+    /**
+     * Returns the utilization according to the state of a queue.
+     * @param available Number of objects available on the queue waiting for being used
+     * @param queueLength Number of pending takes waiting to have access to an object
+     * @param objects Number of objects created on the queue
+     * @return The queue utilization:
+     * - <b>0</b> means all the objects are available and there is no pending takes
+     * - <b>1</b> means all the objects are in use (0 available)
+     * - <b>>1</b> means all the objects are in use (0 available) and there are pending takes
+     */
+    double getUtilization(int available, int queueLength, int objects) {
+        if(objects==0 && queueLength==0) {
+            return 0;
+        }
+        return 1.0 - ((available - queueLength) / Math.max(1.0, objects));
+    }
+
+    /**
+     * Adjust the various queues according to {@link io.aleph.dirigiste.IPool.Controller#adjustment(java.util.Map)}.
+     * Queues no longer in use are removed from the queues and shutdown.
+     */
+    private void adjust() {
+        // The stats need to be accurate at that point to ensure we are not removing a queue
+        // that might be in use. We need to perform a sampling while having the exclusive lock.
+        _lock.lock();
+        sample();
```

Comment:
Since we always need to sample now, why not acquire the lock in the control loop and sample, regardless of whether it's the control period?

Reply:
It was already the case previously.
We were sampling during every iteration.

The only difference is that we are locking during the control period to ensure the computed
utilization is accurate.

Reply:
While we may only adjust during the control period, we do so based on different percentiles, right? So sampling data from a non-control period iteration will affect the adjustment. Most likely, it _won't_ be the data from the control period that ends up in the percentile of interest.

Reply:
> While we may only adjust during the control period, we do so based on different percentiles, right? So sampling data from a non-control period iteration will affect the adjustment. Most likely, it won't be the data from the control period that ends up in the percentile of interest.

What you mention is absolutely right.
On most situation, that's unlikely it will be the value from the control period that will be
the one from interest.
Expect for this situation (let's consider 5 iterations before the control period) 

[0 0 0 0 __0__] => The last iteration will be the value of interest (yes, some of the earlier values can be wrong).

If the last value is not accurate and returns 0 while we are currently trying to create an object on the pool, we might end up with an Exception on the thread that creates the object.

It's totally fine for me to lock when sampling if you think it's necessary.

However:
- It's not the case on master (and might cause too much retention?)
- We are not locking when performing the adjustments (neither on master nor my current branch).

Reply:
https://github.com/clj-commons/dirigiste/pull/32/files#diff-22c0b00c8a29c94adcaba3e9812d82260a3c83d1da581174c4207d3f256652e9R333-R359

Reply:
> [0 0 0 0 0] => The last iteration will be the value of interest (yes, some of the earlier values can be wrong).

I don't understand what you mean by "the last iteration will be the value of interest" here. Because of sorting, the only way the utilization value of interest can be zero is if the whole time window is zeros throughout. Any non-zeros in the reservoir mean the 100th percentile cannot be zero. There's no way for the last iteration being zero to be special. Or is the point that it *shouldn't* be zero but *is*? I'm not following you here.

One reason to lock for all sampling is to avoid any calculation errors that produce non-zero values when not locked, if they _should_ be computing zeros. I don't know for sure there's a problem, but my point above is that if locking to ensure consistency is important, then it must be important for *all* iterations, not just the control period iteration.

> It's not the case on master

Yes, but we already know master has memory leaks 😄 

---

> We are not locking when performing the adjustments

I used to think that would be fine, because that only happens on the `startControlLoop` thread, and dropping a queue requires the pool lock. All the Queue fns use the pool lock (put, take, drop, and shutdown), so we're good there.

However, higher up, neither `acquire()`, nor `queue()` use the lock. This means there's a gap between obtaining the queue reference and adding to it, in the middle of which, the queue could have been removed by adjustment. It would look like it succeeded, too, since an object would be returned to the user, but the queue memory would be out of reach, since `_queues` would no longer hold a reference to it. (GC might still find it when the object is released/disposed, though.) 

Also, it might throw off the numbers of the replacement queue, since `release/dispose` invoke `queue(key)`, so they'd release the object back to a _new_ queue, which will throw off numbers that are part of the Queue itself. Without an initial `take`, `incoming` will be off by one. And if we had to add an object to the old queue via `addObject`, the `objects` count will be one less, because it was incremented on the old queue, not the new one.

Handy Mermaid diagram, because I wanted to try making one 😉 : 

```mermaid
sequenceDiagram
participant Aleph as Aleph thread
participant OldQ as Queue
participant Lock as Pool lock
participant CL as Control loop thread

Note over Aleph,OldQ: Lock not used

%% CL->>OldQ: holds reference in _queues
Aleph->>+OldQ: queue(key) in acquire()
CL->>Lock: .lock()
activate Lock
alt when utilization is 0
    CL->>OldQ: .shutdown()
    CL-->>CL: _queues.remove(key)
    OldQ-->>OldQ: .drop() all objects
end
CL->>Lock: .unlock()
deactivate Lock

Note over Aleph,OldQ: Pool no longer holds a ref to the queue!
Aleph->>OldQ: .addObject()
```

Hmm, Mermaid doesn't support non-nested activations, so I can't easily show how Aleph still holds a reference, but the Pool doesn't. Oh well.

I think we need to lock on `acquire`.

Also, I'm starting to _really_ hate Dirigiste.

Reply:
> I don't understand what you mean by "the last iteration will be the value of interest" here. Because of sorting, the only way the utilization value of interest can be zero is if the whole time window is zeros throughout. Any non-zeros in the reservoir mean the 100th percentile cannot be zero. There's no way for the last iteration being zero to be special. Or is the point that it shouldn't be zero but is? I'm not following you here.

Sorry for not being clear here.
My point is whether all the previous values are 0, we need to ensure the last sampling is correct (0, 1, 2) otherwise we might drop the queue while actually being in use.

So this, yes:

> Or is the point that it shouldn't be zero but is? I'm not following you here.

---

> One reason to lock for all sampling is to avoid any calculation errors that produce non-zero values when not locked, if they should be computing zeros. I don't know for sure there's a problem, but my point above is that if locking to ensure consistency is important, then it must be important for all iterations, not just the control period iteration.

Understood.
My current understanding is that it is not a problem to exhibit inconsistent values outside of the control period.
We might not totally adjust correctly, but is it really a problem?...
I totally get your point though.

----

> I think we need to lock on acquire.

It's late here, but that's exactly why I introduced `_lockedQueues`. To be able to lock on a specific queue on `acquire` and provide a ligther way to lock it without getting the exclusive lock.

(still editing...)

Reply:
Ahh, right, right `_lockedQueues` prevents that. Damn, this is hard to keep track of.

I think I get what you're saying about getting the last utilization val correct, but that leads in to...

> My current understanding is that it is not a problem to exhibit inconsistent values outside of the control period.
We might not totally adjust correctly, but is it really a problem?...

Generally, no, it's not a problem. If utilization is .81 instead of .82, it's not a big deal. Maybe the pool is slightly too big or too small, but that's not the issue. 

My problem scenario is: what if we _consistently_ compute a non-zero utilization when we don't lock for an unused queue, somewhere in the time window? If it was intermittent, we might take longer to remove the queue, but it would still happen. But if we regularly keep getting it wrong, the queue may never get cleared.

The worst case scenario is we get a wrong non-zero utilization under some load 100% of the time. At less than 100%, which is more realistic (I hope), the p of any non-zero in the time window will be 1 - p(ZeroUtilization)<sup>NumSamplesPerWindow</sup>, which will drag it out for a long time, potentially hours/days, but will get cleaned eventually.

Still, I think we can proceed. Even if we get this wrong, I don't think it'll be any worse than before.

Reply:
> My problem scenario is: what if we consistently compute a non-zero utilization when we don't lock for an unused queue, somewhere in the time window? If it was intermittent, we might take longer to remove the queue, but it would still happen. But if we regularly keep getting it wrong, the queue may never get cleared.

Agreed.
However, as long as there is any kind of activity on the queue, we don't really mind if it's not deleted IMO.

Now we are both on the same page, its remains one question:
- This or this? https://github.com/clj-commons/dirigiste/commit/ffafb0db16f1fd34f9931ab33a5f0e7137eb57b4

I still consider that the initial approach was better than the new one where we can still remove a queue while an object is on creation...

I will wait for your confirmation and merge (squash + commit) the PR afterwards.

Reply:
> However, as long as there is any kind of activity on the queue, we don't really mind if it's not deleted IMO

I meant fake non-zero utilization. Like it's not being used, but we're still computing non-zeros somewhere, so it looks like it is.